### PR TITLE
Add base xDS REST SD and kuma_sd implementation

### PR DESCRIPTION
**Pull request by austince**
Implementation of a [Kuma](kuma.io) Service Discovery mechanism via the [xDS REST protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#rest-json-polling-subscriptions).

Since this is new code, the protobuf library used is the official `google.golang.com/protobuf` lib, which is where the community would like to migrate to, as per [this ML thread](https://groups.google.com/g/prometheus-developers/c/uFWRyqZaQis/m/Bi28HO4pAgAJ). The usage is entirely isolated from the `prompb` package where the `gogo/proto` package is currently used. When this transition is complete in Prometheus, it may make sense to include the Kuma MADS v1 `.proto` and compile it here, but since this is a "frozen" API it did not seem too bad to include the compiled version here in the meantime.

In upstream PR, will close:
https://github.com/prometheus/prometheus/issues/7919

## Testing

1. Download and unzip the testing files [kuma-sd-testing-files.zip](https://github.com/austince/prometheus/files/6423696/kuma-sd-testing-files.zip)
2. [Install Kuma](https://kuma.io/docs/1.1.2/installation/helm/) via helm and the `mads-kuma-values.yaml`, which uses images built from `kumahq/kuma:master@97ebe14a3`
2.1. `helm install --namespace kuma-system kuma kuma/kuma -f mads-kuma-values.yaml --create-namespace` 
3. Install the Kuma demo w/ `./mads-demo-install.sh`
4. Install Prometheus via the `mads-prom-metrics.yaml` which will use the `austince/prometheus:kuma-xds-sd` image
5. Generate some load w/ `./mads-demo-gen-load.sh`
6. Port forward Prometheus and see scraped metrics
6.1. `kubectl -n kuma-metrics port-forward svc/prometheus-server 9090:80`

### Screenshots

Screen Shot 2021-05-04 at 4 42 09 PM
Target synchronization

Screen Shot 2021-05-04 at 4 41 44 PM
Targets in Prometheus

Screen Shot 2021-05-04 at 4 43 21 PM
Scraped Envoy metrics

### CODE REVIEW

File: discovery/refresh/refresh.go [hunk 1]
```diff
@@ -115,8 +116,19 @@ func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
...
 	defer d.duration.Observe(time.Since(now).Seconds())
 	tgs, err := d.refreshf(ctx)
-	if err != nil {
+	if err != nil && !errors.Is(err, ErrSkipUpdate) {
 		d.failures.Inc()
 	}
 	return tgs, err
 }
+
+func (d *Discovery) handleError(ctx context.Context, err error) {
```
Comment (lahabana) [inline discovery/refresh/refresh.go:125]: Maybe this could move inside the `refresh()` method?

**Review by lahabana — COMMENTED** [review]
This is very cool!

I was surprised none of the existing usages of Refresher were caching targets.

File: discovery/refresh/refresh.go (hunk 1 above)
Comment (austince) [reply discovery/refresh/refresh.go:125]: Oh yeah, nice!

**Comment by austince** [comment]
Sent PR upstream

**Pull request closed without merging**